src/Symfony/Component/Security/Core/Authentication/Provider/SimpleAuthenticationProvider.php

```diff
@@ -45,6 +46,11 @@ public function authenticate(TokenInterface $token)
         }
 
         $user = $authToken->getUser();
+
+        if (!$user instanceof UserInterface) {
```

Comment:
What if the user is `Anon.`? It will fail

Reply:
No it will not. `instanceof` handles strings.

Reply:
it will call this: `$user = $this->userProvider->loadUserByUsername('Anon.');`, which will fail to retrieve the user and result in `NULL`

Reply:
Yea, but `loadUserByUsername()` will throw a UsernameNotFoundException. Should we throw a specific exception before?

Reply:
It will also handle `null`: https://3v4l.org/Xrt8u

Reply:
@leofeyer instanceof is fine, we wonder about a misleanding username-not-found error in case of an anonymous token, not sure how it should behave yet

Reply:
You cannot guarantee an exception, because the interface doesn't enforce the return type. This means that the provider can return `NULL` instead of an exception:

```
    /**
     * Loads the user for the given username.
     *
     * This method must throw UsernameNotFoundException if the user is not
     * found.
     *
     * @param string $username The username
     *
     * @return UserInterface
     *
     * @throws UsernameNotFoundException if the user is not found
     */
    public function loadUserByUsername($username);
```

Reply:
@chalasr I see, sorry.

Reply:
well, if they don't respect the contract of the interface, they enter unknown land. Things might break badly at any time (and the typehint error on the next call is such a failure), and the failure can even change in any Symfony patch release. We don't guarantee any backward compatibility for code not respecting the interface contract (and yes, any new API added in Symfony 4.1+ will use return types to enforce it better)

Reply:
While I fully agree, sadly (especially older implementations) are often based of "reverse engineered" authenticators, due to lack of better (SF2.0~2.6). Thus, I can't fully blame developers for making this "mistake".

https://github.com/symfony/symfony/blob/9f1c017983a281ad4f29b8681fda1530d4e750c1/src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php#L74-L78

Reply:
same check added for consistency